drivers/dma/dw-cdma/dw-cdma-platform.c

```diff
@@ -0,0 +1,664 @@
+
+	/* The completed descriptor currently is in the head of vc list */
+	vd = vchan_next_desc(&chan->vc);
+	if (unlikely(!vd)) {
+		spin_unlock_irqrestore(&chan->vc.lock, flags);
+		dev_warn(chan2dev(chan), "WARN. Completed descriptor list is empty!\n");
+		return;
+	}
+	desc = vd_to_cdma_desc(vd);
+
+	list_for_each_entry(hw_desc, &desc->hw_desc_issued_list, list) {
+		if (hw_desc->handle == handle) {
+			list_move_tail(&hw_desc->list, &desc->hw_desc_finished_list);
+			entry_notfound = 0;
+			break;
+		}
+	}
+
+	if (unlikely(entry_notfound))
+		dev_warn(chan2dev(chan), "WARN. Didn't find an issued descriptor to complete.");
+
+	done_flg = list_empty(&desc->hw_desc_alloc_list) && list_empty(&desc->hw_desc_issued_list);
+	if (done_flg) {
+		list_del(&vd->node);
+		vchan_cookie_complete(vd);
+	}
+
+	/* Submit queued descriptors after processing the completed ones */
+	cdma_chan_start_first_queued(chan);
+
+	spin_unlock_irqrestore(&chan->vc.lock, flags);
+}
+
+static enum dma_status cdma_chan_tx_status(struct dma_chan *dchan, dma_cookie_t cookie,
+		  struct dma_tx_state *txstate)
+{
+	return dma_cookie_status(dchan, cookie, txstate);
+}
+
+static void process_error_chan(struct c_dma_chan *chan)
+{
+	struct virt_dma_desc *vd;
+	struct c_dma_desc *desc;
+	unsigned long flags;
+
+	spin_lock_irqsave(&chan->vc.lock, flags);
+
+	cdma_reset_ch(chan->id);
+
+	/* The bad descriptor currently is in the head of vc list */
+	vd = vchan_next_desc(&chan->vc);
+	/* Remove the completed descriptor from issued list */
+	list_del(&vd->node);
+
+	desc = vd_to_cdma_desc(vd);
+	list_splice_tail(&desc->hw_desc_alloc_list, &desc->hw_desc_finished_list);
+	list_splice_tail(&desc->hw_desc_issued_list, &desc->hw_desc_finished_list);
+
+	/* WARN about bad descriptor */
+	dev_warn(chan2dev(chan),
+		"WARN. Bad descriptor submitted for cookie: %d\n", vd->tx.cookie);
+
+	vchan_cookie_complete(vd);
+
+	/* Try to restart the controller */
+	cdma_chan_start_first_queued(chan);
+
+	spin_unlock_irqrestore(&chan->vc.lock, flags);
+}
+
+static noinline void cdma_chan_handle_err(struct dw_cluster_dma *dw, u32 err_handle)
+{
+	struct c_dma_chan *chan, *err_chan = NULL;
+	int ii;
+
+	if (err_handle != -1) {
+		/* Handle bus error */
+		for (ii = 0; ii < dw->nr_channels; ii++) {
+			chan = &dw->channels[ii];
+			if (err_handle < chan->desc_base_n || err_handle > chan->desc_last_n)
+				continue;
+			dev_warn(chan2dev(chan), "ERROR. Bus error.\n");
+			process_error_chan(chan);
+			err_chan = chan;
+			break;
+		}
+		if (err_chan == NULL) {
+			for (ii = 0; ii < dw->nr_channels; ii++) {
+				chan = &dw->channels[ii];
+				dev_warn(chan2dev(chan), "ERROR. Bus error.\n");
+				process_error_chan(chan);
+			}
+		}
+	} else {
+		/* Handle overflow.
+		 * Since we don't know from HW which channel caused overflow,
+		 * terminate them all
+		 */
+		for (ii = 0; ii < dw->nr_channels; ii++) {
+			chan = &dw->channels[ii];
+			dev_warn(chan2dev(chan), "WARN. Channel is overflowed.\n");
```

Comment:
probably we can remove 'WARN' from comment - this info is provided by dev_warn. Same for dev_err